SwrveSDK/src/amazon/java/com/swrve/sdk/adm/SwrveAdmIntentService.java

```diff
@@ -0,0 +1,254 @@
+        return !SwrveHelper.isNullOrEmpty(msgId);
+    }
+
+    private void processRemoteNotification(Bundle msg) {
+        if (!isSwrveRemoteNotification(msg)) {
+            SwrveLogger.i(TAG, "ADM notification: but not processing as it doesn't contain: " + SwrveAdmConstants.SWRVE_TRACKING_KEY);
+            return;
+        }
+
+        //Get the context
+        Context context = getApplicationContext();
+
+        //Deduplicate notification
+        //Get tracking key
+        Object rawId = msg.get(SwrveAdmConstants.SWRVE_TRACKING_KEY);
+        String msgId = (rawId != null) ? rawId.toString() : null;
+
+        final String timestamp = msg.getString(SwrveAdmConstants.TIMESTAMP_KEY);
+        if (SwrveHelper.isNullOrEmpty(timestamp)) {
+            SwrveLogger.e(TAG, "ADM notification: but not processing as it's missing " + SwrveAdmConstants.TIMESTAMP_KEY);
+            return;
+        }
+
+        //Check for duplicates. This is a necessary part of using ADM which might clone
+        //a message as part of attempting to deliver it. We de-dupe by
+        //checking against the tracking id and timestamp. (Multiple pushes with the same
+        //tracking id are possible in some scenarios from Swrve).
+        //Id is concatenation of tracking key and timestamp "$_p:$_s.t"
+        String curId = msgId + ":" + timestamp;
+        LinkedList<String> recentIds = getRecentNotificationIdCache(context);
+        if (recentIds.contains(curId)) {
+            //Found a duplicate
+            SwrveLogger.i(TAG, "ADM notification: but not processing because duplicate Id: " + curId);
+            return;
+        }
+
+        //No duplicate found. Update the cache.
+        updateRecentNotificationIdCache(recentIds, curId, MAX_ID_CACHE_ITEMS, context);
+
+        // Notify bound clients
+        Iterator<SwrveQAUser> iter = SwrveQAUser.getBindedListeners().iterator();
+        while (iter.hasNext()) {
+            SwrveQAUser sdkListener = iter.next();
+            sdkListener.pushNotification(msgId, msg);
+        }
+
+        //Process
+        processNotification(msg, context);
+    }
+
+    private LinkedList<String> getRecentNotificationIdCache(Context context) {
+        SharedPreferences sharedPreferences = context.getSharedPreferences(AMAZON_PREFERENCES, Context.MODE_PRIVATE);
+        String jsonString = sharedPreferences.getString(AMAZON_RECENT_PUSH_IDS, "");
+        Gson gson = new Gson();
+        LinkedList<String> recentIds = gson.fromJson(jsonString, new TypeToken<LinkedList<String>>() {}.getType());
+        recentIds = recentIds == null ? new LinkedList<String>() : recentIds;
+        return recentIds;
+    }
+
+    private void updateRecentNotificationIdCache(LinkedList<String> recentIds, String newId, int maxCacheItems, Context context) {
+        //Update queue
+        recentIds.add(newId);
+        //Maintain cache size limit
+        while (recentIds.size() > MAX_ID_CACHE_ITEMS) {
+            recentIds.remove();
```

Comment:
Linked list in java does fifo with add (to end) / remove (from front). toJson writes linked list as a json array (implicitly ordered) and maintaining the order of the linked list. gson.fromJson(test, new TypeToken<LinkedList<String>>() {}.getType()) reads the array into linked list correctly in order too.